### ISSUE TITLE
Grafana 7.2.0 - failed to check table existence: x509: certificate is not valid for any names (Google Cloud SQL)

### ISSUE THREAD

**Issue by dgdevops** [grafana/grafana#27760]
**What happened**:
After upgrading Grafana from 7.1.5 to 7.2.0 the pod get stuck in CrashLoopBackOff due to the following error:
msg="Server shutdown" logger=server reason="Service init failed: migration failed: failed to check table existence: x509: certificate is not valid for any names, but wanted to match <my_sql_instane's_common_name>".

**What you expected to happen**:
After the upgrade Grafana should have come up without any issues as there was no change in any of the configuration besides the image tag change.

**How to reproduce it (as minimally and precisely as possible)**:
1) Use HELM chart (in my case v4.0.0) to deploy Grafana with version 7.1.5
2) Create a MySQL instance to be the backend for Grafana and configure Grafana to use it
3) Enable only secure connections to the SQL instance and pass the server-ca, client-key & client-cert to Grafana
4) Do not modify anything in the HELM values.yaml except the image tag 7.1.5 > 7.2.0

**Anything else we need to know?**:
In case I downgrade back to Grafana 7.1.5 the issue gets resolved perfectly and the pods come up without any issues.

**Environment**:
- Grafana version: 7.2.0
- Data source type & version: MySQL 5.7
- OS Grafana is installed on: Default 7.2.0 image (alpine based)
- User OS & Browser: Windows 10, Chrome
- Grafana plugins: N/A
- Others: Kubernetes 1.16.13-gke.1

**Comment by papagian** [comment]
thank you for reporting this! I have some questions:
mysql is running in an isolated container?
in grafana configuration have you specified the database host IP or the domain?

**Comment by papagian** [comment]
Recently we have upgraded to Go 1.15 which has introduced [X.509 CommonName deprecation](https://golang.org/doc/go1.15#commonname)
can you try setting the environmental variable GODEBUG like:
`GODEBUG=x509ignoreCN=0`
does this, by any chance, resolves your issue?

**Comment by dgdevops** [comment]
Hello @papagian,
Many thanks for your quick response.
1) The MySQL is a GCP managed instance
2) In the Grafana configuration I have the IP of the instance specified

I will perform a test with the GODEBUG environment variable then I will get back to you.

Thank you

**Comment by dgdevops** [comment]
Hello @papagian,
I have added the requested ENV variable to the grafana deployment and the original issue disappeared, however the grafana pod is constantly crashing. I can see the following errors in the logs:

```
t=2020-09-25T10:43:19+0000 lvl=info msg="Starting Grafana" logger=server version=7.2.0 commit=efe4941ee3 branch=HEAD compiled=2020-09-23T11:23:49+0000
t=2020-09-25T10:43:19+0000 lvl=info msg="Config loaded from" logger=settings file=/usr/share/grafana/conf/defaults.ini
t=2020-09-25T10:43:19+0000 lvl=info msg="Config loaded from" logger=settings file=/etc/grafana/grafana.ini
t=2020-09-25T10:43:19+0000 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.data=/var/lib/grafana"
t=2020-09-25T10:43:19+0000 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.logs=/var/log/grafana"
t=2020-09-25T10:43:19+0000 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.plugins=/var/lib/grafana/plugins"
t=2020-09-25T10:43:19+0000 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.provisioning=/etc/grafana/provisioning"
t=2020-09-25T10:43:19+0000 lvl=info msg="Config overridden from command line" logger=settings arg="default.log.mode=console"
t=2020-09-25T10:43:19+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_PATHS_DATA=/var/lib/grafana"
t=2020-09-25T10:43:19+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_PATHS_LOGS=/var/log/grafana"
t=2020-09-25T10:43:19+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_PATHS_PLUGINS=/var/lib/grafana/plugins"
t=2020-09-25T10:43:19+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_PATHS_PROVISIONING=/etc/grafana/provisioning"
t=2020-09-25T10:43:19+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_SECURITY_ADMIN_USER=admin"
t=2020-09-25T10:43:19+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_SECURITY_ADMIN_PASSWORD=*********"
t=2020-09-25T10:43:19+0000 lvl=info msg="Path Home" logger=settings path=/usr/share/grafana
t=2020-09-25T10:43:19+0000 lvl=info msg="Path Data" logger=settings path=/var/lib/grafana
t=2020-09-25T10:43:19+0000 lvl=info msg="Path Logs" logger=settings path=/var/log/grafana
t=2020-09-25T10:43:19+0000 lvl=info msg="Path Plugins" logger=settings path=/var/lib/grafana/plugins
t=2020-09-25T10:43:19+0000 lvl=info msg="Path Provisioning" logger=settings path=/etc/grafana/provisioning
t=2020-09-25T10:43:19+0000 lvl=info msg="App mode production" logger=settings
t=2020-09-25T10:43:19+0000 lvl=info msg="Connecting to DB" logger=sqlstore dbtype=mysql
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x148 pc=0x63e4c1]

goroutine 1 [running]:
crypto/tls.(*Config).Clone(0x0, 0x0)
        /usr/local/go/src/crypto/tls/common.go:733 +0x41
github.com/go-sql-driver/mysql.getTLSConfigClone(0xc0000cf66b, 0x6, 0xc0000cf616)
        /root/go/pkg/mod/github.com/go-sql-driver/mysql@v1.5.0/utils.go:85 +0xcb
github.com/go-sql-driver/mysql.(*Config).normalize(0xc0008e84e0, 0xc0000cf630, 0x41)
        /root/go/pkg/mod/github.com/go-sql-driver/mysql@v1.5.0/dsn.go:130 +0x2d4
github.com/go-sql-driver/mysql.ParseDSN(0xc0000cf600, 0x71, 0x0, 0x7f5b380079a0, 0x5)
        /root/go/pkg/mod/github.com/go-sql-driver/mysql@v1.5.0/dsn.go:361 +0x2bd
github.com/go-sql-driver/mysql.MySQLDriver.OpenConnector(0xc0000cf600, 0x71, 0x1deb860, 0x1dd7bc0, 0xc000149501, 0x7f5b380079a0)
        /root/go/pkg/mod/github.com/go-sql-driver/mysql@v1.5.0/driver.go:100 +0x39
database/sql.Open(0xc000469be1, 0x5, 0xc0000cf600, 0x71, 0xc0008e3968, 0x868869, 0x24a3de0)
        /usr/local/go/src/database/sql/sql.go:771 +0x112
xorm.io/core.Open(0xc000469be1, 0x5, 0xc0000cf600, 0x71, 0x71, 0xc00033fe00, 0x0)
        /root/go/pkg/mod/xorm.io/core@v0.7.3/db.go:86 +0x53
xorm.io/xorm.NewEngine(0xc000469be1, 0x5, 0xc0000cf600, 0x71, 0x2, 0x2, 0xc0008e3ab0)
        /root/go/pkg/mod/xorm.io/xorm@v0.8.1/xorm.go:78 +0x13f
github.com/grafana/grafana/pkg/services/sqlstore.(*SqlStore).getEngine(0xc000251540, 0x248d0a0, 0xc000131640, 0x7f5b37fecd00)
        /root/project/pkg/services/sqlstore/sqlstore.go:299 +0x1b1
github.com/grafana/grafana/pkg/services/sqlstore.(*SqlStore).Init(0xc000251540, 0x2447400, 0xc000251540)
        /root/project/pkg/services/sqlstore/sqlstore.go:71 +0x16d
github.com/grafana/grafana/pkg/server.(*Server).init(0xc00033ecc0, 0x0, 0x0, 0x0)
        /root/project/pkg/server/server.go:145 +0x1f8
github.com/grafana/grafana/pkg/server.(*Server).Run(0xc00033ecc0, 0x0, 0x0)
        /root/project/pkg/server/server.go:156 +0x66
main.main()
        /root/project/pkg/cmd/grafana-server/main.go:125 +0x749
```

Thank you for your help in advance

**Comment by dgdevops** [comment]
Hello @papagian,
Quick update, I downgraded back to 7.1.5 to avoid the SIGSEGV error, however going back to 7.1.5 did not fix it.
As a result the constant grafana pod crash is still happening:

```
t=2020-09-29T08:17:14+0000 lvl=info msg="Starting Grafana" logger=server version=7.1.5 commit=9893b8c53d branch=HEAD compiled=2020-08-25T08:27:17+0000
t=2020-09-29T08:17:14+0000 lvl=info msg="Config loaded from" logger=settings file=/usr/share/grafana/conf/defaults.ini
t=2020-09-29T08:17:14+0000 lvl=info msg="Config loaded from" logger=settings file=/etc/grafana/grafana.ini
t=2020-09-29T08:17:14+0000 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.data=/var/lib/grafana"
t=2020-09-29T08:17:14+0000 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.logs=/var/log/grafana"
t=2020-09-29T08:17:14+0000 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.plugins=/var/lib/grafana/plugins"
t=2020-09-29T08:17:14+0000 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.provisioning=/etc/grafana/provisioning"
t=2020-09-29T08:17:14+0000 lvl=info msg="Config overridden from command line" logger=settings arg="default.log.mode=console"
t=2020-09-29T08:17:14+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_PATHS_DATA=/var/lib/grafana"
t=2020-09-29T08:17:14+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_PATHS_LOGS=/var/log/grafana"
t=2020-09-29T08:17:14+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_PATHS_PLUGINS=/var/lib/grafana/plugins"
t=2020-09-29T08:17:14+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_PATHS_PROVISIONING=/etc/grafana/provisioning"
t=2020-09-29T08:17:14+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_SECURITY_ADMIN_USER=admin"
t=2020-09-29T08:17:14+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_SECURITY_ADMIN_PASSWORD=*********"
t=2020-09-29T08:17:14+0000 lvl=info msg="Path Home" logger=settings path=/usr/share/grafana
t=2020-09-29T08:17:14+0000 lvl=info msg="Path Data" logger=settings path=/var/lib/grafana
t=2020-09-29T08:17:14+0000 lvl=info msg="Path Logs" logger=settings path=/var/log/grafana
t=2020-09-29T08:17:14+0000 lvl=info msg="Path Plugins" logger=settings path=/var/lib/grafana/plugins
t=2020-09-29T08:17:14+0000 lvl=info msg="Path Provisioning" logger=settings path=/etc/grafana/provisioning
t=2020-09-29T08:17:14+0000 lvl=info msg="App mode production" logger=settings
t=2020-09-29T08:17:14+0000 lvl=info msg="Connecting to DB" logger=sqlstore dbtype=mysql
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x62f2b6]
```

**Comment by dgdevops** [comment]
Hello @papagian,
Let me give you a quick update, I managed to fix the SIGSEGV error, my pods with version 7.1.5 came up fine after the downgrade. I did another test with image 7.2.0-ubuntu, also added the GODEBUG environment variable, but the same error is observed:
`msg="Server shutdown" logger=server reason="Service init failed: migration failed: failed to check table existence: x509: certificate is not valid for any names, but wanted to match <my_sql_instane's_common_name>".`

Environmental variable check:
```
/usr/share/grafana $ echo $GODEBUG
x509ignoreCN=0
```

**Comment by papagian** [comment]
thank you for the update

I think that for fixing it we have to set `VerifyConnection` as described [here](https://github.com/golang/go/issues/40748#issuecomment-673612108).

what is the format of your `server_cert_name` configuration? is it something like: `<gcp-project-id>:<cloud-sql-instance>`?

**Comment by dgdevops** [comment]
Hello @papagian,
Thank you for the update.
The `server_cert_name` in the Grafana configuration follows the naming you shared above.
The issue you sent is completely matching my scenario.

1. Is there any workaround that could be used until it is fixed?
2. Are there any plans for fixing this in the next release?

Thank you

**Comment by papagian** [comment]
hello again,
from the discussion in the linked Go issue I understand that this is resolved in the cloudsql-proxy. So as a workaround you could try running cloudsql-proxy and [configure grafana to talk to the proxy instead](https://cloud.google.com/sql/docs/mysql/connect-external-app). Does this help?

If this won't work we can consider including a workaround fix in grafana (unless this will be addressed in Go soon).

**Comment by dgdevops** [comment]
Hello @papagian,

Thank you for your response.
I have tested the cloudsql-proxy workaround and I managed to bring Grafana up by connecting it to the SQL instance on 127.0.0.1:3306. However when I had `ssl_mode = true` enabled I got the following error message:
`lvl=eror msg="Server shutdown" logger=server reason="Service init failed: migration failed: failed to check table existence: TLS requested but server does not support TLS"`

To sum up, the cloudsql-proxy sidecar workaround works but it requires to disable SSL mode in Grafana's configuration.
Additionally, port 3307 needs to be allowed on the firewall, otherwise `Connection refused` error message is observed.

**Comment by marefr** [comment]
I think the idea is that you configure TLS in the cloudsql proxy and connect from Grafana to the proxy (127.0.0.1:3306) without TLS.

> Accessing your Cloud SQL instance using the Cloud SQL Proxy offers the following advantages:
> - Secure connections: The proxy automatically encrypts traffic to and from the database using TLS 1.2 with a 128-bit AES cipher; SSL certificates are used to verify client and server identities.
> - Easier connection management: The proxy handles authentication with Cloud SQL, removing the need to provide static IP addresses.

**Comment by dgdevops** [comment]
Hello @marefr,
Thank you for your response.
From now on, will it be the best practice to use Grafana (from v7.2.0) with cloudsql-proxy or is it a workaround until there will be a fix in Go (if there are actually plans to fix it)?

As many users (including me) are using HELM chart to deploy Grafana the CloudSQL-Proxy sidecar option should be available in the chart to make it easier to deploy the proxy sidecar (in case the usage of the proxy will be required).
Here is an unofficial chart that supports CloudSQL-Proxy sidecar:
https://github.com/mikeraimondi/charts/tree/grafana-cloudsql/stable/grafana.

As I can see a merge attempt was made using the code from the above mentioned repository to the helm:master here:
https://github.com/helm/charts/pull/5831/files

**Comment by marefr** [comment]
Right now cloudsql-proxy would be the solution. But we'll keep this issue open and if more users reporting this problem/upvotes this issue/comments etc we'll consider to work on a fix in Grafana,

**Comment by arslanm** [comment]
I've had the same issue when updating from 7.1.5 to 7.2.1 -- rolled back to 7.1.5.

**Comment by konstantin-kornienko** [comment]
Same problem for AWS RDS.
with ssl_mode:

* true - we have error: "_Server shutdown" logger=server reason="Service init failed: migration failed: failed to check table existence: x509: certificate failed: migration failed: failed to check table existence: x509: certificate relies on legacy Common Name field"_

* skip-verify - everything is working

* fluu - error is: _migration failed: migration failed: failed to check table existence: Error 1045: Access denied for user 'grafana'@_
